ecosystem/sep-0040.md

```diff
@@ -0,0 +1,212 @@
+
+##### 1. URL Storage
+There are two **required** fields on the account itself:
+
+| Field  | Purpose |
+| ------ | ------- |
+| `url`  | A [URL][url] pointing to the "first hop" of metadata about the NFT |
+| `hash` | A [SHA-256](https://datatracker.ietf.org/doc/html/rfc4634#section-4.1) digest of the metadata referenced by the `url` |
+
+The `hash` is provided to ensure integrity of the metadata that the account is pointing to. This prevents the data at the destination from being modified without explicit knowledge and approval by the issuer.
+
+**Warning**: If the hash does not match the metadata content, you should carefully consider the NFT's authenticity.
+
+##### 2. IPFS Content ID (Legacy Method)
+Existing ecosystem solutions lean heavily on IPFS to store NFT content. While this standard aims to free up the ecosystem from dependence on IPFS, it should also work with these existing, organic conventions.
+
+For the legacy variant, there is only one **required** field on the account itself:
+
+| Field      | Purpose |
+| ---------- | ------- |
+| `ipfshash` | An [IPFS Content IDentifier][cid] corresponding to the location of the metadata |
+
+#### IPFS Storage
+If you are issuing a new token and plan to use IPFS, you should use [Method 1](#url-storage) and use an IPFS URL (`ipfs://...`) rather than the legacy method.
+
+If you are the issuer and decide to use IPFS, make sure you understand its implications. You must ensure that the resource is addressable and accessible on the IPFS network by running your own IPFS node or entrusting a provider to do so. Even to look up the NFT, users must use an IPFS lookup service to actually resolve the metadata. For example, [this URL](https://cloudflare-ipfs.com/ipfs/QmaydGEshbL7HMQkD6FZPcMeEt6o15aC4p4P82VxQ7amjy) resolves to the Meridian 2021 Hat NFT via Cloudflare's IPFS node.
+
+#### Standard Metadata Specification
+This section describes the metadata itself rather than how to find it.
+
+The metadata must be valid JSON and must contain the following **required** fields:
+
+| Field       | Description |
+| ----------- | ----------- |
+| `name` | a short name or title for the NFT |
+| `description` | a longer, more-detailed description of the NFT |
+| `type` | a [Media Type][mediatype] describing the NFT's primary format (pointed to by `url`) |
+| `url`  | a [URL][url] pointing to the actual NFT contents |
+| `issuer` | a Stellar public key ([`G...`](https://stellar.org/protocol/sep-23)) corresponding to the account which issued this NFT |
+| `code` | the NFT's asset code |
+
+For example,
+
+```json
+{
+  "name": "A demonstration of NFT metadata",
+  "description": "This is a description of the cool NFT used in this demo.",
+  "type": "image/png",
+  "url": "https://cloudflare-ipfs.com/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/I/m/RickAstleyNeverGonnaGiveYouUp7InchSingleCover.jpg",
+  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
+  "code": "DEMOASSET"
+}
```

Comment:
Could we add the hash of the asset in the metadata?

Reply:
Good call, but I'll echo what I said [above](): I still don't think we should have an enumeration-based approach for hashing. We should either have a type (so `type: sha256` and then `sha256: <hash>`), a standard hashing algorithm under `hash:`, or some short whitelist of hashes (like `sha256: <hash>`, `blake3: <hash>`, etc.)

Reply:
> Could we add the hash of the asset in the metadata?

If I understand what you are proposing @leighmcculloch, I think you would lose the major benefit of having a cryptographic commitment to the content on-chain. The hash that commits to the metadata file must be on-chain. A malicious hoster of the metadata file could change the content of the metadata (including an embedded hash) to entirely new content.  This is similar to the problem that exists today with ERC721 were images hosted and linked to from an NFT could be swapped out or removed with the holder of the NFT not being able to detect that change.

> I still don't think we should have an enumeration-based approach for hashing. 

I would concur with this opinion.  I think there is a lot to be said for the simplicity of a `hash` field that contains arbitrary hash bytes, and a `type` or `hashType` field that contains the description of what the `hash` bytes represent. This is the scheme we are taking at Truestamp.

With this scheme there is always a consistent property name to find what you are looking for. It was mentioned that having a JSON key per hash type allows for multiple hashes to be associated with an NFT, but is this a real use-case? Under which circumstances would a developer need to store both a SHA2-256 and BLAKE3 hash of the same metadata file?

This same scheme can then be applied to having a `data` array at the top level of the metadata file that contains an array of objects representing at least one, but potentially many, image, video, document, etc objects that the metadata points to.  The metadata should also then have a `hash` and `hashType` field for each of those content objects that commits to the hash of the contents of each file. The same `hashType` does not have to be used for each content object in the `data` array.

This affords a simplicity for the developer who can always just look for a single `hash` and `hashType` value at both the on-chain and metadata levels to verify that the metadata file, and all content objects it links to, are probably unchanged..

Reply:
> If I understand what you are proposing

I interpreted this as Leigh wanting an _additional_ hash of the file contents :thinking: Otherwise, I completely agree with you re: integrity on-chain. You could have the issuer sign the metadata and put that in the data itself, but that seems overly convoluted. We just need integrity, since the chain itself provides authenticity.

> I think there is a lot to be said for the simplicity of a hash field that contains arbitrary hash bytes, and a type or hashType field that contains the description of what the hash bytes represent. 

Is there any reason why we wouldn't just stick to a particular hashing algorithm? Every data entry saved is a win in our book.

I really liked the `data` array approach, but after Fred's comments (see below in the issue), it's a little unclear to me where and how we're going to describe NFTs in general now.

Reply:
I was referring to adding a `sha256` field to the metadata JSON that contained a hash of the asset `url` file, assuming that the url is stored somewhere that isn't intrinsically verifiable like IPFS is.

Reply:
I agree this discussion is mostly not relevant if we are exploring a focus on SEP-1.